src/exml_stream.erl

```diff
@@ -60,131 +45,45 @@ new_parser() ->
-        StartTag = proplists:get_value(start_tag, Opts),
-        {ok, EventParser} =
-            case StartTag of
-                undefined -> exml_event:new_parser(MaxChildSize);
-                _ -> exml_event:new_parser(MaxChildSize, StartTag)
-            end,
-
-        {ok, #parser{
-            event_parser = EventParser,
-            config = #config{
-                        infinite_stream = bool_opt(infinite_stream, Opts, false),
-                        autoreset = bool_opt(autoreset, Opts, false),
-                        start_tag = StartTag
-                        }}}
-    catch
-        E:R -> {error, {E, R}}
-    end.
+    MaxChildSize = proplists:get_value(max_child_size, Opts, 0),
+    InfiniteStream = proplists:get_value(infinite_stream, Opts, false),
+    {ok, EventParser} = exml_nif:create(MaxChildSize, InfiniteStream),
+    {ok, #parser{event_parser = EventParser, buffer = <<>>}}.
 
 -spec parse(parser(), binary()) -> {ok, parser(), [exml_stream:element()]} | {error, Reason :: any()}.
-parse(Parser, Input) ->
-    #parser{event_parser = EventParser, stack = OldStack, config = Config} = Parser,
-    #config{autoreset = Autoreset, infinite_stream = InfiniteStream, start_tag = StartTag} = Config,
-    case exml_event:parse(EventParser, Input) of
-        {error, Reason} -> {error, Reason};
-        {ok, Events} ->
-            case parse_events(Events, OldStack, InfiniteStream, StartTag) of
-                {error, Reason} -> {error, Reason};
-                {[], Elements} when Autoreset =:= true ->
-                    {ok, NewParser} = reset_parser(Parser),
-                    {ok, NewParser, lists:reverse(Elements)};
-                {NewStack, Elements} ->
-                    {ok, Parser#parser{stack = NewStack}, lists:reverse(Elements)}
-            end
+parse(Parser, Input) when is_binary(Input) ->
+    #parser{event_parser = EventParser, buffer = OldBuf} = Parser,
+    Buffer = <<OldBuf/binary, Input/binary>>,
+    case parse_all(EventParser, Buffer, []) of
+        {ok, Elems, RestBuffer} ->
+            {ok, Parser#parser{buffer = RestBuffer}, Elems};
+        Other ->
+            Other
     end.
 
--spec reset_parser(parser()) -> {ok, parser()} | {error, any()}.
-reset_parser(#parser{event_parser = EventParser, config = Config}) ->
-    try
-        exml_event:reset_parser(EventParser),
-        %% drop all the state except event_parser
-        {ok, #parser{event_parser = EventParser, config = Config}}
-    catch
-        E:R ->
-            {error, {E, R}}
-    end.
+-spec reset_parser(parser()) -> {ok, parser()}.
+reset_parser(#parser{event_parser = NifParser} = Parser) ->
+    exml_nif:reset_parser(NifParser),
+    {ok, Parser#parser{buffer = <<>>}}.
 
--spec free_parser(parser()) -> ok | {error, any()}.
-free_parser(#parser{event_parser = EventParser}) ->
-    exml_event:free_parser(EventParser).
+-spec free_parser(parser()) -> ok.
+free_parser(#parser{}) ->
+    ok.
```

Comment:
The parser is a resource with a destructor, it will be garbage collected like other Erlang terms: https://github.com/esl/exml/pull/32/files/930a1ba739372da3868429b13f2abe93d95cfda7#diff-86dd88dbbd595f450931f8b97ae56831R429

Reply:
:+1: